fil-ast/src/expr.rs

```diff
@@ -1,5 +1,170 @@
 use super::{Binding, Id, Loc};
 use fil_utils::Error;
+use itertools::Itertools;
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+/// A concrete value for a parameter, internally either a float or an integer
+pub enum Concrete {
+    Float(f64),
+    UInt(u64),
+}
+
+impl Concrete {
+    /// Convert a [Concrete] value into a uint
+    pub fn u64(self) -> u64 {
+        self.try_into().unwrap_or_else(|_| {
+            unreachable!("Value {} could not be converted into a uint.", self)
+        })
+    }
+
+    /// Convert a [Concrete] value into a float
+    pub fn f64(self) -> f64 {
+        self.try_into().unwrap_or_else(|_| {
+            unreachable!("Value {} could not be converted into a float.", self)
+        })
+    }
+}
+
+impl std::fmt::Display for Concrete {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Concrete::Float(n) => write!(f, "{}", n),
+            Concrete::UInt(n) => write!(f, "{}", n),
+        }
+    }
+}
+
+impl From<u64> for Concrete {
+    fn from(v: u64) -> Self {
+        Concrete::UInt(v)
+    }
+}
+
+impl From<f64> for Concrete {
+    fn from(v: f64) -> Self {
+        Concrete::Float(v)
+    }
+}
+
+impl TryFrom<Concrete> for u64 {
+    type Error = Error;
+
+    fn try_from(value: Concrete) -> Result<Self, Self::Error> {
+        match value {
+            Concrete::UInt(n) => Ok(n),
+            Concrete::Float(_) => Err(Error::malformed(
+                "Cannot concretize float value into uint.",
+            )),
+        }
+    }
+}
+
+impl TryFrom<Concrete> for f64 {
+    type Error = Error;
+
+    fn try_from(value: Concrete) -> Result<Self, Self::Error> {
+        match value {
+            Concrete::UInt(n) => Ok(n as f64),
```

Comment:
@rachitnigam We should discuss whether uints auto-casting to floats should be allowed. This simplifies implementation pretty significantly because functions can take floats and they'll automatically evaluate for integers as well, but it would be (insignificantly) lossy for very large `u64`s

Reply:
FWIW, this shouldn't change behavior at all currently, as internally `u64` concrete values were already cast to `f64` to perform `log` (and that is legitimately probably the best way to do it).